server/router/route_test.go

```diff
@@ -232,3 +243,231 @@ func isMessageReceived(route *Route, msg *protocol.Message) bool {
+	})
+
+	msMock.EXPECT().MaxMessageID("fetch_request").Return(uint64(2), nil).Times(2)
+
+	routerMock.EXPECT().Done().Return(make(chan bool)).AnyTimes()
+	routerMock.EXPECT().Fetch(gomock.Any()).Do(func(req *store.FetchRequest) {
+		a.Equal(req.Partition, "fetch_request")
+		a.Equal(uint64(0), req.StartID)
+		a.Equal(uint64(0), req.EndID)
+		a.Equal(store.DirectionForward, req.Direction)
+
+		// send to messages
+		req.Push(1, []byte(strings.Replace(dummyMessageBytes, "MESSAGE_ID", strconv.Itoa(1), 1)))
+		req.Push(2, []byte(strings.Replace(dummyMessageBytes, "MESSAGE_ID", strconv.Itoa(2), 1)))
+		req.Done()
+	})
+
+	done := make(chan struct{})
+	go func() {
+		receivedMessages := 0
+		for i := 1; i <= 2; i++ {
+			select {
+			case m, opened := <-route.MessagesChannel():
+				if opened {
+					receivedMessages++
+					a.Equal(uint64(i), m.ID)
+				}
+			case <-time.After(50 * time.Millisecond):
+				a.Fail("Message not received")
+			}
+		}
+		a.Equal(2, receivedMessages)
+		close(done)
+	}()
+
+	err := route.Provide(routerMock, false)
+	a.NoError(err)
+	<-done
+}
+
+func TestRoute_Provide_WithSubscribe(t *testing.T) {
+	ctrl, finish := testutil.NewMockCtrl(t)
+	defer finish()
+
+	a := assert.New(t)
+	msMock := NewMockMessageStore(ctrl)
+	routerMock := NewMockRouter(ctrl)
+
+	routerMock.EXPECT().MessageStore().Return(msMock, nil)
+
+	route := NewRoute(RouteConfig{
+		Path:         protocol.Path("/fetch_request"),
+		ChannelSize:  4,
+		FetchRequest: store.NewFetchRequest("", 0, 0, store.DirectionForward, -1),
+	})
+
+	routerMock.EXPECT().Done().Return(make(chan bool)).AnyTimes()
+	routerMock.EXPECT().Fetch(gomock.Any()).Do(func(req *store.FetchRequest) {
+		a.Equal(req.Partition, "fetch_request")
+		a.Equal(uint64(0), req.StartID)
+		a.Equal(uint64(0), req.EndID)
+		a.Equal(store.DirectionForward, req.Direction)
+
+		// send to messages
+		req.Push(1, []byte(strings.Replace(dummyMessageBytes, "MESSAGE_ID", strconv.Itoa(1), 1)))
+		req.Push(2, []byte(strings.Replace(dummyMessageBytes, "MESSAGE_ID", strconv.Itoa(2), 1)))
+		req.Done()
+	})
+
+	msMock.EXPECT().MaxMessageID(gomock.Eq("fetch_request")).Return(uint64(2), nil).Times(2)
+
+	routerMock.EXPECT().Subscribe(gomock.Any()).Do(func(r *Route) (*Route, error) {
+		a.Equal(route, r)
+
+		for i := 3; i <= 4; i++ {
+			r.Deliver(&protocol.Message{
+				ID:   uint64(i),
+				Path: "/fetch_request",
+				Body: []byte("dummy"),
+			})
+		}
+
+		return r, nil
+	})
+
+	done := make(chan struct{})
+	go func() {
+		receivedMessages := 0
+		for i := 1; i <= 4; i++ {
+			select {
+			case m, opened := <-route.MessagesChannel():
+				if opened {
+					receivedMessages++
+					a.Equal(uint64(i), m.ID)
+				}
+			case <-time.After(50 * time.Millisecond):
+				a.Fail("Message not received")
+			}
+		}
+		a.Equal(4, receivedMessages)
+		close(done)
+	}()
+
+	err := route.Provide(routerMock, true)
+	a.NoError(err)
+	<-done
+}
+
+type startable interface {
```

Comment:
do we need to extract these 2 interfaces (from router and service) in a separate package, and make them public ?